### PR TITLE
chore: upgrade DataFusion family, again

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Upgrade DataFusion to the commit just before using Arrow 58; and sqlparser to v0.61.0

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [ ] API changes are backward compatible.
- [ ] Schema or data changes are backward compatible.

### CODE REVIEW

**Comment by MichaelScofield** [comment]
@codex review

**Comment by MichaelScofield** [comment]
@gemini-code-assist review

File: tests/cases/standalone/common/types/decimal/decimal_ops.result [hunk 1]
```diff
@@ -233,15 +225,15 @@ SELECT ROUND('999.9'::DECIMAL(4,1)), ROUND('99999999.9'::DECIMAL(9,1)), ROUND('9
 +----------------------+---------------------------+------------------------------------+--------------------------------------------------------+
 | round(Utf8("999.9")) | round(Utf8("99999999.9")) | round(Utf8("99999999999999999.9")) | round(Utf8("9999999999999999999999999999999999999.9")) |
 +----------------------+---------------------------+------------------------------------+--------------------------------------------------------+
-| 1000.0               | 100000000.0               | 1e17                               | 1e37                                                   |
+| 100.0                | 10000000.0                | 10000000000000000.0                | 1000000000000000000000000000000000000.0                |
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply tests/cases/standalone/common/types/decimal/decimal_ops.result]: wait for https://github.com/apache/datafusion/issues/19921

**Comment by sunng87** [comment]
Any chance we can merge this on a develop branch to move on, if the round fix doesn't have an ETA?

**Comment by MichaelScofield** [comment]
> Any chance we can merge this on a develop branch to move on, if the round fix doesn't have an ETA?

round fix is done, that's why I redo this pr

**Comment by MichaelScofield** [comment]
@codex review again

**Comment by MichaelScofield** [comment]
@gemini-code-assist review again

File: Cargo.toml [hunk 2]
```diff
@@ -331,21 +331,22 @@ git = "https://github.com/GreptimeTeam/greptime-meter.git"
...
+datafusion-functions = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-functions-aggregate-common = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-functions-window-common = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-optimizer = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-physical-expr = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-physical-expr-common = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-physical-plan = { git = "https://github.com/GreptimeTeam/datafusion.git", rev = "02b82535e0160c4545667f36a03e1ff9d1d2e51f" }
+datafusion-pg-catalog = { git = "https://github.com/GreptimeTeam/datafusion-postgres.git", rev = "284cb351c02210dc60162b0ec70a71d35f33803d" }
```
Comment (sunng87) [inline Cargo.toml]: We can use released version of datafusion-pg-catalog if our datafusion is 52

File: tests/cases/distributed/optimizer/filter_push_down.result [hunk 3]
```diff
@@ -110,7 +110,6 @@ SELECT * FROM integers WHERE i NOT IN ((SELECT i FROM integers WHERE i=1)) ORDER
 +---+-------------------------+
 | 2 | 1970-01-01T00:00:00.002 |
 | 3 | 1970-01-01T00:00:00.003 |
-|   | 1970-01-01T00:00:00.004 |
```
Comment (sunng87) [inline tests/cases/distributed/optimizer/filter_push_down.result:113]: double check this

File: tests/cases/standalone/common/aggregate/approx_percentile_cont_with_weight.result [hunk 4]
```diff
@@ -16,7 +16,7 @@ SELECT approx_percentile_cont_with_weight(weight, 0.5) WITHIN GROUP (ORDER BY "v
 +---------------------------------------------------------------------------------------------------------------------+
 | approx_percentile_cont_with_weight(weight_test.weight,Float64(0.5)) WITHIN GROUP [weight_test.value ASC NULLS LAST] |
 +---------------------------------------------------------------------------------------------------------------------+
-| 22                                                                                                                  |
+| 33                                                                                                                  |
```
Comment (sunng87) [inline tests/cases/standalone/common/aggregate/approx_percentile_cont_with_weight.result:19]: double check

File: tests/cases/standalone/common/flow/flow_basic.result [hunk 5]
```diff
@@ -1218,8 +1218,8 @@ FROM
 | stat | bucket_size | total_logs | time_window         |
 +------+-------------+------------+---------------------+
 | 200  | 100         | 1          | 1970-01-01T00:00:00 |
-| 200  | 200         | 1          | 1970-01-01T00:00:00 |
-| 200  | 210         | 3          | 1970-01-01T00:00:00 |
+| 200  | 200         | 3          | 1970-01-01T00:00:00 |
+| 200  | 210         | 1          | 1970-01-01T00:00:00 |
```
Comment (sunng87) [inline tests/cases/standalone/common/flow/flow_basic.result:1222]: double check

File: tests/cases/standalone/common/types/decimal/decimal_ops.result [hunk 6]
```diff
@@ -191,40 +187,36 @@ SELECT FLOOR('999.9'::DECIMAL(4,1)), FLOOR('99999999.9'::DECIMAL(9,1)), FLOOR('9
...
 +----------------------+---------------------------+------------------------------------+--------------------------------------------------------+
 
 SELECT FLOOR('-999.9'::DECIMAL(4,1)), FLOOR('-99999999.9'::DECIMAL(9,1)), FLOOR('-99999999999999999.9'::DECIMAL(18,1)), FLOOR('-9999999999999999999999999999999999999.9'::DECIMAL(38,1));
 
-+-----------------------+----------------------------+-------------------------------------+---------------------------------------------------------+
-| floor(Utf8("-999.9")) | floor(Utf8("-99999999.9")) | floor(Utf8("-99999999999999999.9")) | floor(Utf8("-9999999999999999999999999999999999999.9")) |
-+-----------------------+----------------------------+-------------------------------------+---------------------------------------------------------+
-| -1000.0               | -100000000.0               | -1e17                               | -1e37                                                   |
-+-----------------------+----------------------------+-------------------------------------+---------------------------------------------------------+
+Error: 3001(EngineExecuteQuery), Compute error: Decimal overflow while applying floor
```
Comment (sunng87) [inline tests/cases/standalone/common/types/decimal/decimal_ops.result:195]: behaviour changes

File: src/query/src/log_query/planner.rs [hunk 7]
```diff
@@ -964,18 +964,18 @@ mod tests {
             exprs: vec![LogExpr::ScalarFunc {
                 name: "date_trunc".to_string(),
                 args: vec![
-                    LogExpr::NamedIdent("timestamp".to_string()),
                     LogExpr::Literal("day".to_string()),
+                    LogExpr::NamedIdent("timestamp".to_string()),
```
Comment (fengys1996) [inline src/query/src/log_query/planner.rs:968]: nitpick: Is reordering necessary?

File: Cargo.toml (hunk 2 above)
Comment (MichaelScofield) [reply Cargo.toml]: not yet, some codes need to be changed: https://github.com/GreptimeTeam/datafusion-postgres/tree/adapt-new-sqlparser

File: tests/cases/distributed/optimizer/filter_push_down.result (hunk 3 above)
Comment (MichaelScofield) [reply tests/cases/distributed/optimizer/filter_push_down.result:113]: This is the correct result. Value compared to `NULL` is "UNKNOWN" in SQL, see [this stackoverflow post](https://stackoverflow.com/questions/129077/null-values-inside-not-in-clause). It's fixed in https://github.com/apache/datafusion/pull/19635

File: tests/cases/standalone/common/aggregate/approx_percentile_cont_with_weight.result (hunk 4 above)
Comment (MichaelScofield) [reply tests/cases/standalone/common/aggregate/approx_percentile_cont_with_weight.result:19]: We don't implement these "weighted approximate percentile" functions by our own. This test is actually testing DataFusion's implementation. I believe now is the correct result, related DataFusion change is https://github.com/apache/datafusion/pull/19941

File: tests/cases/standalone/common/flow/flow_basic.result (hunk 5 above)
Comment (MichaelScofield) [reply tests/cases/standalone/common/flow/flow_basic.result:1222]: I think now is the correct result, given what I see from the above inserted data on table "ngx_access_log". 

"bucket" with size 200 does have 3 logs:

```
("cli1", 200, 200, 1),
("cli1", 200, 205, 1),
("cli1", 200, 209, 1),
```

while with size 210, there is only 1:

```
("cli1", 200, 210, 1),
```

Can you verify this? @discord9

File: tests/cases/standalone/common/types/decimal/decimal_ops.result (hunk 6 above)
Comment (MichaelScofield) [reply tests/cases/standalone/common/types/decimal/decimal_ops.result:195]: Another DataFusion regression like the `ROUND` function on the decimal values. But unlike the `ROUND`, which produced incorrect results, the `FLOOR` returns an error. Acceptable to some degree. Let's not make this a blocker to this PR, it's waited too long. I'll submit an issue to DataFusion.

File: Cargo.toml (hunk 2 above)
Comment (sunng87) [reply Cargo.toml]: I see. Let me try if I can include this in mainline

Comment (sunng87) [reply Cargo.toml]: https://github.com/datafusion-contrib/datafusion-postgres/pull/301

Will create a release when it's ready.

File: src/promql/src/extension_plan/normalize.rs [hunk 8]
```diff
@@ -273,6 +273,10 @@ impl ExecutionPlan for SeriesNormalizeExec {
     }
 
     fn required_input_distribution(&self) -> Vec<Distribution> {
+        if self.tag_columns.is_empty() {
```
Comment (discord9) [inline src/promql/src/extension_plan/normalize.rs:276]: This seems reasonable, but could it cause reduce of parallelism in some case or maybe some other corner cases? @waynexia might want to take a look?

File: tests/cases/standalone/common/flow/flow_basic.result (hunk 5 above)
Comment (discord9) [reply tests/cases/standalone/common/flow/flow_basic.result:1222]: > I think now is the correct result, given what I see from the above inserted data on table "ngx_access_log".
> 
> "bucket" with size 200 does have 3 logs:
> 
> ```
> ("cli1", 200, 200, 1),
> ("cli1", 200, 205, 1),
> ("cli1", 200, 209, 1),
> ```
> 
> while with size 210, there is only 1:
> 
> ```
> ("cli1", 200, 210, 1),
> ```
> 
> Can you verify this? [@discord9](https://github.com/discord9)

it seems trunc(number, -1) 's behavior changed from `floor` to rounding

Comment (discord9) [reply tests/cases/standalone/common/flow/flow_basic.result:1222]: > I think now is the correct result, given what I see from the above inserted data on table "ngx_access_log".
> 
> "bucket" with size 200 does have 3 logs:
> 
> ```
> ("cli1", 200, 200, 1),
> ("cli1", 200, 205, 1),
> ("cli1", 200, 209, 1),
> ```
> 
> while with size 210, there is only 1:
> 
> ```
> ("cli1", 200, 210, 1),
> ```
> 
> Can you verify this? @discord9

Ok it seems  previsouly we are in a bugged datafusion version, see https://github.com/apache/datafusion/pull/19794 which fixed trunc from being `round` to properly trunc

File: tests/cases/standalone/common/types/interval/interval.result [hunk 9]
```diff
@@ -71,11 +71,6 @@ SELECT arrow_typeof(INTERVAL '1 month');
 | Interval(MonthDayNano)                                                                            |
 +---------------------------------------------------------------------------------------------------+
 
--- INTERVAL + TIME CONSTANT
-SELECT current_time() + INTERVAL '1 hour';
-
-Error: 3000(PlanQuery), Failed to plan SQL: Error during planning: Cannot coerce arithmetic expression Time64(ns) + Interval(MonthDayNano) to valid types
```
Comment (discord9) [inline tests/cases/standalone/common/types/interval/interval.result:77]: seems datafusion now allowed this, this wouldn't break anything right?

File: tests/cases/standalone/optimizer/filter_push_down.result [hunk 10]
```diff
@@ -98,7 +98,6 @@ SELECT * FROM integers WHERE i NOT IN ((SELECT i FROM integers WHERE i=1)) ORDER
 +---+-------------------------+
 | 2 | 1970-01-01T00:00:00.002 |
 | 3 | 1970-01-01T00:00:00.003 |
-|   | 1970-01-01T00:00:00.004 |
```
Comment (discord9) [inline tests/cases/standalone/optimizer/filter_push_down.result:101]: Related fix is this: https://github.com/apache/datafusion/pull/19635 so it's correct behavior now

**Review by discord9 — APPROVED** [review]
rest LGTM

File: tests/cases/standalone/common/types/interval/interval.result (hunk 9 above)
Comment (MichaelScofield) [reply tests/cases/standalone/common/types/interval/interval.result:77]: Right, the reason I remove it is because `current_time()` makes Sqlness test result unstable.

File: src/query/src/log_query/planner.rs (hunk 7 above)
Comment (MichaelScofield) [reply src/query/src/log_query/planner.rs:968]: Yes, I remember it's because the function's signature has changed.
